lib/msal-common/src/crypto/PopTokenGenerator.ts

```diff
@@ -37,36 +37,45 @@ export class PopTokenGenerator {
     }
 
     async generateCnf(request: BaseAuthRequest): Promise<string> {
+        const reqCnf = await this.generateKid(request);
+        return this.cryptoUtils.base64Encode(JSON.stringify(reqCnf));
+    }
+
+    async generateKid(request: BaseAuthRequest): Promise<ReqCnf> {
         const kidThumbprint = await this.cryptoUtils.getPublicKeyThumbprint(request);
-        const reqCnf: ReqCnf = {
+
+        return {
             kid: kidThumbprint,
             xms_ksl: KeyLocation.SW
         };
-        return this.cryptoUtils.base64Encode(JSON.stringify(reqCnf));
     }
 
     async signPopToken(accessToken: string, request: BaseAuthRequest): Promise<string> {
         const tokenClaims: TokenClaims | null = AuthToken.extractTokenClaims(accessToken, this.cryptoUtils);
+        if (!tokenClaims?.cnf?.kid) {
+            throw ClientAuthError.createTokenClaimsRequiredError();
+        }
+        
+        return this.signPayload(accessToken, tokenClaims.cnf.kid, request);
+    }
 
+    async signPayload(payload: string, kid: string, request: SignedHttpRequestParameters, claims?: object): Promise<string> {
         // Deconstruct request to extract SHR parameters
         const { resourceRequestMethod, resourceRequestUri, shrClaims } = request;
 
         const resourceUrlString = (resourceRequestUri) ? new UrlString(resourceRequestUri) : undefined;
         const resourceUrlComponents = resourceUrlString?.getUrlComponents();
 
-        if (!tokenClaims?.cnf?.kid) {
-            throw ClientAuthError.createTokenClaimsRequiredError();
-        }
-
         return await this.cryptoUtils.signJwt({
-            at: accessToken,
+            at: payload,
             ts: TimeUtils.nowSeconds(),
             m: resourceRequestMethod?.toUpperCase(),
             u: resourceUrlComponents?.HostNameAndPort,
             nonce: this.cryptoUtils.createNewGuid(),
             p: resourceUrlComponents?.AbsolutePath,
             q: (resourceUrlComponents?.QueryString) ? [[], resourceUrlComponents.QueryString] : undefined,
-            client_claims: shrClaims || undefined
-        }, tokenClaims.cnf.kid);
+            client_claims: shrClaims || undefined,
+            ...claims
```

Comment:
Do we need to worry about claim-name collisions? One of the reasons shrClaims are nested in their own claim is to avoid ts/m/u/p/q from being accidentally overwritten by custom claims.

Reply:
We specifically want apps to override claims in the SHR, thus spreading the entire object. `nonce` is the main one we care about for now.

Reply:
I guess this brings up another question: if a `nonce` is provided in `client_claims`, does our middleware use that instead?

Reply:
We don't parse `client_claims` by design, it is also treated as an opaque string. I think whatever is passed in `client_claims` is a contract between app and resource server that MSAL has no knowledge of. 

And I see what you mean, given that this class is specifically for advanced use cases in which apps need to control what goes in the SHR, it makes sense to let those claims override the MSAL-generated ones.

Reply:
> Do we need to add a spec item for our middleware to accommodate this?

I don't think so given,

> We don't parse client_claims by design, it is also treated as an opaque string. I think whatever is passed in client_claims is a contract between app and resource server that MSAL has no knowledge of.